src/jit/importer.cpp

```diff
@@ -3583,21 +3583,41 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
 
         case CORINFO_INTRINSIC_GetTypeFromHandle:
             op1 = impStackTop(0).val;
-            if (op1->gtOper == GT_CALL && (op1->gtCall.gtCallType == CT_HELPER) &&
-                gtIsTypeHandleToRuntimeTypeHelper(op1->AsCall()))
+
+            if (IsTargetAbi(CORINFO_CORERT_ABI))
             {
-                op1 = impPopStack().val;
-                // Change call to return RuntimeType directly.
-                op1->gtType = TYP_REF;
-                retNode     = op1;
+                if (op1->gtOper == GT_CALL && (op1->gtCall.gtCallType == CT_HELPER) &&
+                    op1->AsCall()->gtCallMethHnd == eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE))
+                {
+                    op1 = impPopStack().val;
+                    // Replace helper with a more specialized helper that returns RuntimeType
+                    op1 = op1->gtCall.gtCallArgs;
+                    assert(op1->OperIsList());
+                    assert(op1->gtOp.gtOp2 == nullptr);
+                    GenTreeArgList* helperArgs = gtNewArgList(op1->gtOp.gtOp1);
+                    op1 = gtNewHelperCallNode(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE, TYP_REF, helperArgs);
```

Comment:
You should be able to reuse the old arglist here.

Reply:
Just to confirm I understand this... does this sound right?

`CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE` is generally cheaper to invoke as it doesn't require allocating a managed object. But if we think the jit can optimize we're willing to change this helper to the more expensive `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE` which the downstream opts already recognize.

An alternative would be to plumb through understanding of the new helper in the downstream optimization code. This might offer some (small) advantage as sometimes trees don't come together until later on, eg when we propagate inline return values.

Reply:
The problem is really just about RyuJIT currently not seeing the difference between `ldtoken Foo` and `ldtoken Foo / call GetTypeFromHandle`, because both are represented as the result of the same helper (`CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE`) - only the return type of the helper is bashed to TYP_REF for the latter case.

On CoreRT, these two represent different values, so we need something to distinguish between these. `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE` serves this purpose.

I think it should be enough to do this in the importer. The `ldtoken Foo / call GetTypeFromHandle` is a pretty fixed sequence that the C# compiler generates. I don't think we need to optimize for the case when the `ldtoken` comes from an inline.

Reply:
Makes sense, thanks. 

Just to restate:

On CoreCLR we have two types, a pointer (runtime type) and a struct wrapped pointer (runtime type handle), and we rely on the fact that they're treated the same by all our existing ABIs, so the helper can be seen as returning either type, depending on context.

On CoreRT this isn't the case, and so we have to use different helpers.

Reply:
That's correct!